File: python/11_validate_bst.py
"""
Q11. Validate Binary Search Tree (Category: Trees)

Problem: Determine if a binary tree is a valid BST.

Approach: Pass min/max bounds down the recursion. Each node must fall
          strictly within (min, max).

Time Complexity : O(n)
Space Complexity: O(h)
"""
from __future__ import annotations
from typing import Optional
import math


class TreeNode:
    def __init__(self, val: int = 0,
                 left: Optional["TreeNode"] = None,
                 right: Optional["TreeNode"] = None):
        self.val = val
        self.left = left
        self.right = right


def is_valid_bst(root: Optional[TreeNode]) -> bool:
    def validate(node: Optional[TreeNode], low: float, high: float) -> bool:
        if node is None:
            return True
        if not (low < node.val < high):
            return False
        return validate(node.left, low, node.val) and \
               validate(node.right, node.val, high)

    return validate(root, -math.inf, math.inf)


if __name__ == "__main__":
    # Valid BST
    root1 = TreeNode(2, TreeNode(1), TreeNode(3))
    print(is_valid_bst(root1))  # True

    # Invalid BST
    root2 = TreeNode(5, TreeNode(1),
                     TreeNode(4, TreeNode(3), TreeNode(6)))
    print(is_valid_bst(root2))  # False

File: cpp/11_ValidateBST.cpp
/*
 * Q11. Validate Binary Search Tree (Category: Trees)
 *
 * Problem: Determine if a binary tree is a valid BST.
 *
 * Approach: Pass min/max bounds down the recursion. Each node must fall
 *           strictly within (min, max).
 *
 * Time Complexity : O(n)
 * Space Complexity: O(h)
 */
#include <iostream>
#include <climits>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

bool validate(TreeNode* node, long min, long max) {
    if (!node) return true;
    if (node->val <= min || node->val >= max) return false;
    return validate(node->left, min, node->val) &&
           validate(node->right, node->val, max);
}

bool isValidBST(TreeNode* root) {
    return validate(root, (long)INT_MIN - 1, (long)INT_MAX + 1);
}

int main() {
    cout << boolalpha;

    // Valid BST: 2 / 1 / 3
    TreeNode* r1 = new TreeNode(2);
    r1->left  = new TreeNode(1);
    r1->right = new TreeNode(3);
    cout << isValidBST(r1) << endl; // true

    // Invalid BST: 5 / 1 / 4(3,6)
    TreeNode* r2 = new TreeNode(5);
    r2->left  = new TreeNode(1);
    r2->right = new TreeNode(4);
    r2->right->left  = new TreeNode(3);
    r2->right->right = new TreeNode(6);
    cout << isValidBST(r2) << endl; // false
    return 0;
}
